File: go/dijkstra/otimizado/dijkstra_otimizado.go
package main

import (
	"container/heap"
	"fmt"
	"math"
)

// Edge representa uma aresta com peso
type Edge struct {
	to     int
	weight int
}

// WeightedGraph representa um grafo ponderado usando lista de adjacência
type WeightedGraph struct {
	vertices int
	adjList  [][]Edge
}

// NewWeightedGraph cria um novo grafo ponderado
func NewWeightedGraph(vertices int) *WeightedGraph {
	return &WeightedGraph{
		vertices: vertices,
		adjList:  make([][]Edge, vertices),
	}
}

// AddEdge adiciona uma aresta ponderada
func (g *WeightedGraph) AddEdge(from, to, weight int) {
	g.adjList[from] = append(g.adjList[from], Edge{to, weight})
}

// PriorityQueue implementa uma fila de prioridade usando heap
// Otimização: O(log V) para inserção e remoção vs O(V) na versão básica
type PriorityQueue []*Item

type Item struct {
	vertex   int
	distance int
	index    int
}

func (pq PriorityQueue) Len() int { return len(pq) }

func (pq PriorityQueue) Less(i, j int) bool {
	return pq[i].distance < pq[j].distance
}

func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].index = i
	pq[j].index = j
}

func (pq *PriorityQueue) Push(x interface{}) {
	n := len(*pq)
	item := x.(*Item)
	item.index = n
	*pq = append(*pq, item)
}

func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	old[n-1] = nil
	item.index = -1
	*pq = old[0 : n-1]
	return item
}

// Dijkstra encontra os caminhos mais curtos de um vértice para todos os outros
// Complexidade: O((V + E) log V) com heap vs O(V²) na versão básica
func (g *WeightedGraph) Dijkstra(start int) ([]int, []int) {
	const INF = math.MaxInt32

	distances := make([]int, g.vertices)
	previous := make([]int, g.vertices)
	visited := make([]bool, g.vertices)

	// Inicializa distâncias
	for i := 0; i < g.vertices; i++ {
		distances[i] = INF
		previous[i] = -1
	}
	distances[start] = 0

	// Cria fila de prioridade otimizada com heap
	pq := make(PriorityQueue, 0, g.vertices)
	heap.Init(&pq)
	heap.Push(&pq, &Item{vertex: start, distance: 0})

	for pq.Len() > 0 {
		// Extrai o vértice com menor distância (O(log V))
		current := heap.Pop(&pq).(*Item)
		currentVertex := current.vertex

		// Skip se já foi processado (pode haver duplicatas na fila)
		if visited[currentVertex] {
			continue
		}

		visited[currentVertex] = true

		// Relaxa todas as arestas adjacentes
		for _, edge := range g.adjList[currentVertex] {
			neighbor := edge.to
			newDistance := distances[currentVertex] + edge.weight

			// Relaxamento: tenta melhorar a distância
			if newDistance < distances[neighbor] {
				distances[neighbor] = newDistance
				previous[neighbor] = currentVertex
				// Adiciona à fila de prioridade (O(log V))
				heap.Push(&pq, &Item{vertex: neighbor, distance: newDistance})
			}
		}
	}

	return distances, previous
}

// DijkstraOptimized versão que para quando encontra o destino específico
// Útil quando só queremos o caminho para um vértice específico
func (g *WeightedGraph) DijkstraOptimized(start, end int) (int, []int) {
	const INF = math.MaxInt32

	if start == end {
		return 0, []int{start}
	}

	distances := make([]int, g.vertices)
	previous := make([]int, g.vertices)
	visited := make([]bool, g.vertices)

	for i := 0; i < g.vertices; i++ {
		distances[i] = INF
		previous[i] = -1
	}
	distances[start] = 0

	pq := make(PriorityQueue, 0, g.vertices)
	heap.Init(&pq)
	heap.Push(&pq, &Item{vertex: start, distance: 0})

	for pq.Len() > 0 {
		current := heap.Pop(&pq).(*Item)
		currentVertex := current.vertex

		if visited[currentVertex] {
			continue
		}

		visited[currentVertex] = true

		// Otimização: para quando encontra o destino
		if currentVertex == end {
			break
		}

		for _, edge := range g.adjList[currentVertex] {
			neighbor := edge.to
			newDistance := distances[currentVertex] + edge.weight

			if newDistance < distances[neighbor] {
				distances[neighbor] = newDistance
				previous[neighbor] = currentVertex
				heap.Push(&pq, &Item{vertex: neighbor, distance: newDistance})
			}
		}
	}

	path := g.GetPath(previous, start, end)
	return distances[end], path
}

// GetPath reconstrói o caminho de start para end
func (g *WeightedGraph) GetPath(previous []int, start, end int) []int {
	if previous[end] == -1 && start != end {
		return nil // Não há caminho
	}

	path := make([]int, 0)
	current := end

	for current != -1 {
		path = append([]int{current}, path...)
		current = previous[current]
	}

	return path
}

// DisplayGraph exibe a representação da lista de adjacência
func (g *WeightedGraph) DisplayGraph() {
	fmt.Println("Grafo (Lista de Adjacência):")
	for i := 0; i < g.vertices; i++ {
		fmt.Printf("Vértice %d: ", i)
		for _, edge := range g.adjList[i] {
			fmt.Printf("->(%d,%d) ", edge.to, edge.weight)
		}
		fmt.Println()
	}
	fmt.Println()
}

// BenchmarkComparison compara diferentes versões do algoritmo
func (g *WeightedGraph) BenchmarkComparison(start int) {
	fmt.Println("=== Comparação de Performance ===")

	fmt.Println("Executando Dijkstra otimizado...")
	distances, previous := g.Dijkstra(start)

	fmt.Printf("Resultados para vértice %d:\n", start)
	for i := 0; i < g.vertices; i++ {
		if distances[i] != math.MaxInt32 {
			path := g.GetPath(previous, start, i)
			fmt.Printf("  Para %d: distância %d, caminho %v\n", i, distances[i], path)
		} else {
			fmt.Printf("  Para %d: INALCANÇÁVEL\n", i)
		}
	}
}

// Exemplo de uso
func main() {
	fmt.Println("=== Algoritmo de Dijkstra - Versão Otimizada ===")

	// Cria um grafo maior para demonstrar as otimizações
	g := NewWeightedGraph(8)

	// Adiciona arestas para criar um grafo mais complexo
	g.AddEdge(0, 1, 4)
	g.AddEdge(0, 2, 2)
	g.AddEdge(1, 2, 1)
	g.AddEdge(1, 3, 5)
	g.AddEdge(2, 3, 8)
	g.AddEdge(2, 4, 10)
	g.AddEdge(3, 4, 2)
	g.AddEdge(3, 5, 6)
	g.AddEdge(4, 5, 3)
	g.AddEdge(0, 6, 8)
	g.AddEdge(6, 7, 1)
	g.AddEdge(5, 7, 4)

	g.DisplayGraph()

	fmt.Println("1. Dijkstra completo (todos os caminhos):")
	g.BenchmarkComparison(0)

	fmt.Println("\n2. Dijkstra otimizado (caminho específico):")
	distance, path := g.DijkstraOptimized(0, 7)
	fmt.Printf("De 0 para 7: distância = %d, caminho = %v\n", distance, path)

	fmt.Println("\n=== Otimizações Implementadas ===")
	fmt.Println("1. Lista de Adjacência: O(V + E) espaço vs O(V²) matriz")
	fmt.Println("2. Heap (Priority Queue): O(log V) vs O(V) para encontrar mínimo")
	fmt.Println("3. Early Termination: Para quando encontra destino específico")
	fmt.Println("4. Complexidade Total: O((V + E) log V) vs O(V²)")

	fmt.Println("\n=== Quando Usar Cada Versão ===")
	fmt.Println("Básica: Grafos pequenos, fins didáticos, todos os caminhos")
	fmt.Println("Otimizada: Grafos grandes, performance crítica, busca específica")
}

File: python/dijkstra/otimizado/dijkstra_otimizado.py
"""
Implementação Otimizada do Algoritmo de Dijkstra
===============================================

Esta implementação utiliza uma heap (fila de prioridade) para otimizar
a busca pelo vértice com menor distância, resultando em melhor performance.

Características desta versão:
- Usa heapq para fila de prioridade eficiente
- Complexidade: O((V + E) log V) onde V = vértices, E = arestas
- Mais eficiente para grafos grandes
- Parada antecipada quando destino é alcançado

Autor: Algorithms Repository
Data: Julho 2025
"""

import heapq
import sys
from typing import Dict, List, Tuple

class GrafoDijkstraOtimizado:
    """
    Classe que representa um grafo para o algoritmo de Dijkstra otimizado.
    """
    
    def __init__(self):
        """
        Inicializa um grafo vazio.
        """
        self.vertices = {}  # Dicionário de adjacências: {vertice: [(vizinho, peso), ...]}
        
    def adicionar_vertice(self, vertice: str) -> None:
        """
        Adiciona um vértice ao grafo.
        
        Args:
            vertice: Nome do vértice a ser adicionado
        """
        if vertice not in self.vertices:
            self.vertices[vertice] = []
            
    def adicionar_aresta(self, origem: str, destino: str, peso: float) -> None:
        """
        Adiciona uma aresta ponderada ao grafo.
        
        Args:
            origem: Vértice de origem
            destino: Vértice de destino
            peso: Peso da aresta (deve ser não-negativo)
        """
        if peso < 0:
            raise ValueError("O algoritmo de Dijkstra não funciona com pesos negativos")
            
        # Garante que ambos os vértices existem
        self.adicionar_vertice(origem)
        self.adicionar_vertice(destino)
        
        # Adiciona a aresta (grafo direcionado)
        self.vertices[origem].append((destino, peso))
        
    def dijkstra_otimizado(self, origem: str, destino: str) -> Tuple[float, List[str]]:
        """
        Executa o algoritmo de Dijkstra otimizado com heap.
        
        Args:
            origem: Vértice de origem
            destino: Vértice de destino
            
        Returns:
            Tupla contendo (distância_total, caminho)
            
        Raises:
            ValueError: Se origem ou destino não existirem no grafo
        """
        if origem not in self.vertices:
            raise ValueError(f"Vértice de origem '{origem}' não existe no grafo")
        if destino not in self.vertices:
            raise ValueError(f"Vértice de destino '{destino}' não existe no grafo")
            
        # Inicialização
        distancias = {vertice: sys.maxsize for vertice in self.vertices}
        distancias[origem] = 0
        
        # Para reconstruir o caminho
        predecessores = {vertice: None for vertice in self.vertices}
        
        # Heap para fila de prioridade: (distância, vértice)
        heap = [(0, origem)]
        
        # Conjunto de vértices processados
        processados = set()
        
        print(f"Iniciando Dijkstra Otimizado de '{origem}' para '{destino}'")
        print("Usando heap para fila de prioridade")
        
        # Algoritmo principal
        while heap:
            # Obtém o vértice com menor distância
            distancia_atual, atual = heapq.heappop(heap)
            
            # Se já processamos este vértice, ignora
            if atual in processados:
                continue
                
            # Marca como processado
            processados.add(atual)
            
            print(f"\nProcessando vértice: {atual} (distância: {distancia_atual})")
            
            # Se chegamos ao destino, podemos parar (otimização)
            if atual == destino:
                print("Destino alcançado! Parando busca.")
                break
                
            # Relaxa todas as arestas do vértice atual
            for vizinho, peso in self.vertices[atual]:
                if vizinho not in processados:
                    nova_distancia = distancia_atual + peso
                    
                    if nova_distancia < distancias[vizinho]:
                        print(f"  Relaxando {vizinho}: {distancias[vizinho]} -> {nova_distancia}")
                        distancias[vizinho] = nova_distancia
                        predecessores[vizinho] = atual
                        
                        # Adiciona à heap com nova distância
                        heapq.heappush(heap, (nova_distancia, vizinho))
                        
        # Reconstrói o caminho
        caminho = []
        if distancias[destino] != sys.maxsize:
            atual = destino
            while atual is not None:
                caminho.append(atual)
                atual = predecessores[atual]
            caminho.reverse()
            
        return distancias[destino], caminho
        
    def dijkstra_todos_caminhos(self, origem: str) -> Dict[str, Tuple[float, List[str]]]:
        """
        Calcula os caminhos mais curtos da origem para todos os outros vértices.
        
        Args:
            origem: Vértice de origem
            
        Returns:
            Dicionário com {destino: (distância, caminho)} para todos os vértices
        """
        if origem not in self.vertices:
            raise ValueError(f"Vértice de origem '{origem}' não existe no grafo")
            
        # Inicialização
        distancias = {vertice: sys.maxsize for vertice in self.vertices}
        distancias[origem] = 0
        
        # Para reconstruir os caminhos
        predecessores = {vertice: None for vertice in self.vertices}
        
        # Heap para fila de prioridade
        heap = [(0, origem)]
        processados = set()
        
        print(f"Calculando todos os caminhos a partir de '{origem}'")
        
        # Algoritmo principal
        while heap:
            distancia_atual, atual = heapq.heappop(heap)
            
            if atual in processados:
                continue
                
            processados.add(atual)
            
            # Relaxa todas as arestas
            for vizinho, peso in self.vertices[atual]:
                if vizinho not in processados:
                    nova_distancia = distancia_atual + peso
                    
                    if nova_distancia < distancias[vizinho]:
                        distancias[vizinho] = nova_distancia
                        predecessores[vizinho] = atual
                        heapq.heappush(heap, (nova_distancia, vizinho))
                        
        # Reconstrói todos os caminhos
        resultados = {}
        for destino in self.vertices:
            if destino != origem:
                caminho = []
                if distancias[destino] != sys.maxsize:
                    atual = destino
                    while atual is not None:
                        caminho.append(atual)
                        atual = predecessores[atual]
                    caminho.reverse()
                    
                resultados[destino] = (distancias[destino], caminho)
                
        return resultados
        
    def imprimir_grafo(self) -> None:
        """
        Imprime uma representação do grafo.
        """
        print("Representação do Grafo:")
        for vertice in sorted(self.vertices.keys()):
            vizinhos = ", ".join([f"{v}({p})" for v, p in self.vertices[vertice]])
            print(f"  {vertice} -> {vizinhos if vizinhos else 'sem conexões'}")


def exemplo_performance():
    """
    Exemplo demonstrando a performance da versão otimizada.
    """
    print("=== Exemplo de Performance - Dijkstra Otimizado ===\n")
    
    # Cria um grafo maior para demonstrar a eficiência
    grafo = GrafoDijkstraOtimizado()
    
    # Cria uma estrutura de grafo em grade
    tamanho = 5
    for i in range(tamanho):
        for j in range(tamanho):
            vertice = f"v{i}{j}"
            
            # Conecta com vizinhos (direita e baixo)
            if j < tamanho - 1:  # Conecta à direita
                grafo.adicionar_aresta(vertice, f"v{i}{j+1}", 1)
            if i < tamanho - 1:  # Conecta para baixo
                grafo.adicionar_aresta(vertice, f"v{i+1}{j}", 1)
                
    # Adiciona algumas arestas diagonais com peso maior
    for i in range(tamanho - 1):
        for j in range(tamanho - 1):
            grafo.adicionar_aresta(f"v{i}{j}", f"v{i+1}{j+1}", 1.5)
            
    print(f"Grafo criado com {len(grafo.vertices)} vértices")
    
    # Teste de caminho específico
    origem = "v00"
    destino = "v44"
    
    print(f"\nBuscando caminho de {origem} para {destino}:")
    print("=" * 50)
    
    distancia, caminho = grafo.dijkstra_otimizado(origem, destino)
    
    print("\nResultado:")
    if caminho:
        print(f"Caminho: {' -> '.join(caminho)}")
        print(f"Distância total: {distancia}")
    else:
        print("Não há caminho entre os vértices.")


def exemplo_todos_caminhos():
    """
    Exemplo demonstrando o cálculo de todos os caminhos.
    """
    print("\n" + "="*60)
    print("=== Exemplo - Todos os Caminhos ===")
    print("="*60 + "\n")
    
    # Cria um grafo pequeno para demonstrar
    grafo = GrafoDijkstraOtimizado()
    
    # Adiciona um hub central
    grafo.adicionar_aresta("Hub", "A", 2)
    grafo.adicionar_aresta("Hub", "B", 1)
    grafo.adicionar_aresta("Hub", "C", 3)
    grafo.adicionar_aresta("A", "D", 1)
    grafo.adicionar_aresta("B", "D", 2)
    grafo.adicionar_aresta("C", "D", 1)
    grafo.adicionar_aresta("A", "E", 4)
    grafo.adicionar_aresta("B", "E", 3)
    grafo.adicionar_aresta("D", "E", 1)
    
    grafo.imprimir_grafo()
    
    # Calcula todos os caminhos a partir do Hub
    origem = "Hub"
    todos_caminhos = grafo.dijkstra_todos_caminhos(origem)
    
    print(f"\nTodos os caminhos mais curtos a partir de '{origem}':")
    print("-" * 50)
    
    for destino, (distancia, caminho) in todos_caminhos.items():
        if caminho:
            print(f"{destino}: {' -> '.join(caminho)} (distância: {distancia})")
        else:
            print(f"{destino}: sem caminho")


def comparacao_algoritmos():
    """
    Demonstra a diferença entre as versões básica e otimizada.
    """
    print("\n" + "="*60)
    print("=== Comparação: Básico vs Otimizado ===")
    print("="*60)
    
    print("\nCaracterísticas da versão BÁSICA:")
    print("- Usa lista simples para encontrar vértice com menor distância")
    print("- Complexidade: O(V²)")
    print("- Mais didática e fácil de entender")
    print("- Processa todos os vértices")
    
    print("\nCaracterísticas da versão OTIMIZADA:")
    print("- Usa heap (fila de prioridade) para eficiência")
    print("- Complexidade: O((V + E) log V)")
    print("- Mais eficiente para grafos grandes")
    print("- Parada antecipada quando destino é encontrado")
    print("- Função adicional para todos os caminhos")
    
    print("\nRecomendação:")
    print("- Use a versão BÁSICA para aprender os conceitos")
    print("- Use a versão OTIMIZADA em aplicações reais")


if __name__ == "__main__":
    exemplo_performance()
    exemplo_todos_caminhos()
    comparacao_algoritmos()
